### PR TITLE
sql: refactor to combine `forEachTableDesc*` functions.

### PR DESCRIPTION
**Pull request by shubhamdhama**
Background:

I wanted to include deleted tables in `crdb_internal.table_spans` and add a
new column `is_deleted`. However, to achieve this, I needed to add a
boolean to `forEachTableDesc` and all the invoked functions so that
`forEachTableDescWithTableLookupInternalFromDescriptors` doesn't skip them.
The argument count for these functions was already excessive, so I made the
following changes to improve readability and maintainability.

Details:

Previously, there were around four different overloads of functions:
- forEachTableDesc
- forEachTableDescAll
- forEachTableDescWithTableLookup
- forEachTableDescAllWithTableLookup
- forEachTableDescWithTableLookupInternal

This added excessive nesting and reduced readability. In many cases,
functions with the prefix `TableLookup` were used even though they no
longer utilized `tableLookupFn`.

In these proposed changes:
- I aim to improve readability by merging these functions.
- To avoid repeatedly declaring the same definition, I moved all arguments
  to a new struct. This way, we only define what we need.

```diff
- func(ctx context.Context, db catalog.DatabaseDescriptor,
- 		 sc catalog.SchemaDescriptor, table catalog.TableDescriptor) error {
+ func(ctx context.Context, descCtx tableDescContext) error {
+     sc, table := descCtx.schema, descCtx.table
```

Now, we only have `forEachTableDesc` to iterate over table descriptors.

Informs: #128578, #97942
Epic: CRDB-12100
Release note: None

### CODE REVIEW

**Comment by shubhamdhama** [comment]
This refactor would allow me to work on my next PR (described in the background). I would really appreciate a review on this and if this whole refactor makes sense or not.

I don't feel very strong about all this, especially because I'm new to this module. I'm okay if we feel this is not the right direction. If that would be the case, I would go with adding a new function or new argument to this long list of arguments (or please suggest some other direction).

And if we feel like this PR helps everyone and is in right direction, I have some todos
- [ ] Double sure all the places are updated correctly. I've tried to make changes in such a way so that it won't produce even bigger diff. 
- [ ] Make sure tests are passing
- [ ] Update code comments

**Comment by shubhamdhama** [comment]
The core of the changes are in https://github.com/cockroachdb/cockroach/pull/128755/files#diff-2dd67f0db19a5dba0340ee2973244c9747b18396ff239bb2e9b67f160ebd92a7L2645-L2865

**Review by rafiss — CHANGES_REQUESTED** [review]
this is a great direction to go in. it is a bit tricky to make sure that all the nested calls were refactored to the new way correctly, but we should have a goal of having zero test differences from before and after this patch. were you able to get all the tests into a passing state?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-:-O5PApfU2ZmnfO3v7GFM:b-zddk3s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shubhamdhama)

___
*[`pkg/sql/information_schema.go` line 2606 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-O5PBEgr2A2sJn_HQd1y:-O5PBEgr2A2sJn_HQd1z:bzgnvng) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fbc4ec06dd0238a08768c222d7353ecd877cc6e/pkg/sql/information_schema.go#L2606)):*
> ```Go
> const (
> 	// hideVirtual completely hides virtual schemas during iteration.
> 	hideVirtual virtualOpts = iota
> ```

nit: why did `hideVirtual` need to become the first enum value? i worry that this could cause unintended differences, since some callers may have been relying on the zero value before.
___
*[`pkg/sql/information_schema.go` line 2667 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-O5PA5fY0Qewvp8tyYCT:-O5PA5fY0Qewvp8tyYCU:b-bzyo9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fbc4ec06dd0238a08768c222d7353ecd877cc6e/pkg/sql/information_schema.go#L2667)):*
> ```Go
> 	lCtx := newInternalLookupCtx(c.OrderedDescriptors(), dbContext)
> 
> 	virtualOpts := opts.virtualOpts
> ```

this is causing static code analysis to fail:
```
pkg/sql/information_schema.go:2667:2: declaration of "virtualOpts" shadows declaration at line 2602 (shadow)
```
___
*[`pkg/sql/pg_catalog.go` line 3602 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-O5PAYe37rmkW1rO9t3X:-O5PAYe37rmkW1rO9t3Y:bjah44q) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fbc4ec06dd0238a08768c222d7353ecd877cc6e/pkg/sql/pg_catalog.go#L3602)):*
> ```Go
> 				if err := forEachTableDesc(ctx, p, dbContext, opts, func(
> 					ctx context.Context, descCtx tableDescContext) error {
> 					db, sc, table := descCtx.database, descCtx.schema, descCtx.table
> ```

this is causing a static analysis issue
```
pkg/sql/pg_catalog.go:3602:6: declaration of "db" shadows declaration at line 3587 (shadow)
```

**Review by shubhamdhama — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128755)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/information_schema.go` line 2606 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-O5PBEgr2A2sJn_HQd1y:-O5SofPRCOKvDH5sjqcU:bqp2ekb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fbc4ec06dd0238a08768c222d7353ecd877cc6e/pkg/sql/information_schema.go#L2606)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: why did `hideVirtual` need to become the first enum value? i worry that this could cause unintended differences, since some callers may have been relying on the zero value before.
</blockquote></details>

Since it was a required argument so it was safe, but I'm reverting this to reduce unnecessary risk.
___
*[`pkg/sql/information_schema.go` line 2667 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-O5PA5fY0Qewvp8tyYCT:-O5S24xA46yL4Zu9zaDm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fbc4ec06dd0238a08768c222d7353ecd877cc6e/pkg/sql/information_schema.go#L2667)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is causing static code analysis to fail:
```
pkg/sql/information_schema.go:2667:2: declaration of "virtualOpts" shadows declaration at line 2602 (shadow)
```
</blockquote></details>

Done.
___
*[`pkg/sql/pg_catalog.go` line 3602 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-O5PAYe37rmkW1rO9t3X:-O5SHSKq3J5ymlu6-0MF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fbc4ec06dd0238a08768c222d7353ecd877cc6e/pkg/sql/pg_catalog.go#L3602)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is causing a static analysis issue
```
pkg/sql/pg_catalog.go:3602:6: declaration of "db" shadows declaration at line 3587 (shadow)
```
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
nice work! the code is easier to read now. thanks for getting this working.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128755#-:-O5TraO07Y3Rbp5Hobgp:b-n8l7xl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

File: pkg/sql/information_schema.go [hunk 1]
```diff
@@ -2651,77 +2610,36 @@ const (
...
-// forEachTableDescWithTableLookup.
 //
-// The allowAdding argument if true includes newly added tables that
-// are not yet public.
-// The validate argument if false turns off checking if the descriptor ids exist
-// and if they are valid.
-func forEachTableDescWithTableLookupInternal(
+// The virtualOpts argument specifies how virtual tables are made
+// visible.
+func forEachTableDesc(
```
Comment (shubhamdhama) [inline pkg/sql/information_schema.go:2637]: Fix function comment.

**Comment by shubhamdhama** [comment]
Thanks for the review!

bors r=rafiss

**Pull request merged**